# Webscraping 

https://br.advfn.com/bolsa-de-valores/bovespa/A

In [1]:
import requests

from bs4 import BeautifulSoup

Documentação  
https://www.crummy.com/software/BeautifulSoup/bs4/doc/

In [2]:
url = 'https://br.advfn.com/bolsa-de-valores/bovespa/A'

page = requests.get(url)

page

<Response [200]>

A mensagem chegou ao servidor e era válida.  
OK (200) - tudo ocorreu exatamente como planejado: a requisição estava OK e o servidor respondeu adequadamente.

In [10]:
url = 'https://br.advfn.com/bolsa-de-valores/bovespa/Affff'

page = requests.get(url)

page

<Response [404]>

A requisição possui algum erro  
Not Found (404) - o recurso buscado não existe.

In [14]:
url = 'https://br.advfn.com/bolsa-de-valores/bovespa/Af'

try:
    page = requests.get(url)
    page.raise_for_status()
except Exception as err:
    print('Algo deu errado:', err)
    

Algo deu errado: 404 Client Error: Not Found for url: https://br.advfn.com/bolsa-de-valores/bovespa/Af


In [19]:
url = 'https://br.advfn.com/bolsa-de-valores/bovespa/A'

try:
    page = requests.get(url)
    page.raise_for_status()
    response = page.text #ou page.content
    soup = BeautifulSoup(response, 'html.parser')
    
except Exception as err:
    print('Algo deu errado:', err)
    


In [20]:
soup

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">

<html lang="pt-BR" xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta content="IE=edge" http-equiv="X-UA-Compatible"/>
<script type="text/javascript">
window.USER_COUNTRY = "BR";
window.USER_STATE = "SP";
window.USER_FROM_EU = "";
</script>
<!-- APS_HEADER_START -->
<script type="text/javascript">
function AdvfnJsQueueLoader(e){this.check=e,this.queue=[],this.id=!1,this.ready=!1;var t=this;this.id=setInterval(function(){t.tick()},50)}AdvfnJsQueueLoader.prototype.processQueue=function(){var e;if(0<this.queue.length)for(e=0;e<this.queue.length;e++)this.queue[e]()},AdvfnJsQueueLoader.prototype.tick=function(){this.check()&&(clearInterval(this.id),this.ready=!0,this.processQueue())},AdvfnJsQueueLoader.prototype.add=function(e){this.ready?e():this.queue.push(e)};
var jqueryReady=new AdvfnJsQueueLoader(function(){return'function'==typeof $});
</script>
<script type="text/javascript">
window.USER_COUNT

In [21]:
soup.title

<title>A - Bovespa: Lista completa de ativos A-Z da Bolsa de Valores do Brasil e Mundo de A.O. Smith a AÇO ALTONA PN</title>

In [22]:
soup.title.name

'title'

In [23]:
soup.title.string

'A - Bovespa: Lista completa de ativos A-Z da Bolsa de Valores do Brasil e Mundo de A.O. Smith a AÇO ALTONA PN'

Para facilitar:  
http://akul.me/blog/2016/beautifulsoup-cheatsheet/

In [24]:
# visualizar mais facilmente o objeto html
print(soup.prettify())

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html lang="pt-BR" xmlns="http://www.w3.org/1999/xhtml">
 <head>
  <meta content="IE=edge" http-equiv="X-UA-Compatible"/>
  <script type="text/javascript">
   window.USER_COUNTRY = "BR";
window.USER_STATE = "SP";
window.USER_FROM_EU = "";
  </script>
  <!-- APS_HEADER_START -->
  <script type="text/javascript">
   function AdvfnJsQueueLoader(e){this.check=e,this.queue=[],this.id=!1,this.ready=!1;var t=this;this.id=setInterval(function(){t.tick()},50)}AdvfnJsQueueLoader.prototype.processQueue=function(){var e;if(0<this.queue.length)for(e=0;e<this.queue.length;e++)this.queue[e]()},AdvfnJsQueueLoader.prototype.tick=function(){this.check()&&(clearInterval(this.id),this.ready=!0,this.processQueue())},AdvfnJsQueueLoader.prototype.add=function(e){this.ready?e():this.queue.push(e)};
var jqueryReady=new AdvfnJsQueueLoader(function(){return'function'==typeof $});
  </script>
  <script type="text/javascript

In [25]:
# encontrar partes do html que iremos trabalhar
soup.find_all('tr') # saída é uma lista das informações que ele encontrou

[<tr class="table_heading"><th class="Column1 String">Ação</th>
 <th class="Column2 ColumnLast String"></th>
 </tr>,
 <tr class="even first"><td class="String Column1"><a href="//br.advfn.com/bolsa-de-valores/bovespa/acaracoacarapna-pna-ACCC5L/cotacao" title="Cotação da ACARACOACARAPNA (ACCC5L)">ACARACOACARAPNA PNA</a></td><td class="String Column2 ColumnLast">ACCC5L</td></tr>,
 <tr class="odd"><td class="String Column1"><a href="//br.advfn.com/bolsa-de-valores/bovespa/advanced-dh-on-ADHM3/cotacao" title="Cotação da Advanced Digital Health Medicina Preventiva SA (ADHM3)">ADVANCED-DH ON</a></td><td class="String Column2 ColumnLast">ADHM3</td></tr>,
 <tr class="even"><td class="String Column1"><a href="//br.advfn.com/bolsa-de-valores/bovespa/advanced-dh-on-ADHM9/cotacao" title="Cotação da Advanced Digital Health Medicina Preventiva SA (ADHM9)">ADVANCED-DH ON</a></td><td class="String Column2 ColumnLast">ADHM9</td></tr>,
 <tr class="odd"><td class="String Column1"><a href="//br.advfn.com/

In [26]:
# len em uma lista, quantidade tr que ele encontrou
len(soup.find_all('tr'))

90

tr td = table row table data

In [27]:
# outros  exemplos
soup.find_all('div')

[<div id="splitWindow" mode="" url1="https://br.advfn.com/bolsa-de-valores/bovespa/A" url2="https://br.advfn.com/bolsa-de-valores/bovespa/A"></div>,
 <div class="normal" id="header-container">
 <div id="header">
 <div id="logo-area">
 <a href="/" id="advfn-logo" title="Cotações de Ações da Bolsa de Valores Bovespa">
 <img alt="ADVFN Brasil" border="0" src="/common/images/logos/br.advfn.com.gif"/>
 </a>
 <span id="clock">
 <span class="sb" id="clock_dt">21/07/2021</span> <span class="sb" id="clock_tm">23:29:13</span>
 </span>
 <span id="support-box">
 <span id="support-email"><a href="/p.php?pid=prodmatrix&amp;xref=header_upgrade_text" id="prod-matrix-br">Produtos ADVFN</a></span>
 <span id="support-phone"><a href="/fale-conosco" id="prod-matrix-br">Fale Conosco</a></span>
 </span>
 <span id="account-buttons">
 <a class="header-btn" href="//br.advfn.com/common/account/register?xref=header_regBtn">Cadastro Gratis</a>
 <a class="header-btn" href="//br.advfn.com/common/account/login" role=

In [28]:
soup.find_all('td')

[<td class="String Column1"><a href="//br.advfn.com/bolsa-de-valores/bovespa/acaracoacarapna-pna-ACCC5L/cotacao" title="Cotação da ACARACOACARAPNA (ACCC5L)">ACARACOACARAPNA PNA</a></td>,
 <td class="String Column2 ColumnLast">ACCC5L</td>,
 <td class="String Column1"><a href="//br.advfn.com/bolsa-de-valores/bovespa/advanced-dh-on-ADHM3/cotacao" title="Cotação da Advanced Digital Health Medicina Preventiva SA (ADHM3)">ADVANCED-DH ON</a></td>,
 <td class="String Column2 ColumnLast">ADHM3</td>,
 <td class="String Column1"><a href="//br.advfn.com/bolsa-de-valores/bovespa/advanced-dh-on-ADHM9/cotacao" title="Cotação da Advanced Digital Health Medicina Preventiva SA (ADHM9)">ADVANCED-DH ON</a></td>,
 <td class="String Column2 ColumnLast">ADHM9</td>,
 <td class="String Column1"><a href="//br.advfn.com/bolsa-de-valores/bovespa/advanced-dh-on-ADHM1/cotacao" title="Cotação da Advanced Digital Health Medicina Preventiva SA (ADHM1)">ADVANCED-DH ON</a></td>,
 <td class="String Column2 ColumnLast">AD

In [29]:
# Mostra todos os parágrafos do texto
soup.find_all('p')

[<p>Siga-nos nas redes sociais</p>,
 <p> Ao acessar os serviços da ADVFN você estará de acordo com os 
                 <a href="/common/legal/terms-and-conditions" title="Termos e Condições">Termos e Condições</a>
 </p>]

In [31]:
# Encontrar a informação que queremos, no caso a lista de ações e seus tickers
soup.find_all(class_='String Column1') 
# Pego a classe 'String Column1' e ele mostra tudo que ele encontrou (nesse caso, o nome das ações)

[<td class="String Column1"><a href="//br.advfn.com/bolsa-de-valores/bovespa/acaracoacarapna-pna-ACCC5L/cotacao" title="Cotação da ACARACOACARAPNA (ACCC5L)">ACARACOACARAPNA PNA</a></td>,
 <td class="String Column1"><a href="//br.advfn.com/bolsa-de-valores/bovespa/advanced-dh-on-ADHM3/cotacao" title="Cotação da Advanced Digital Health Medicina Preventiva SA (ADHM3)">ADVANCED-DH ON</a></td>,
 <td class="String Column1"><a href="//br.advfn.com/bolsa-de-valores/bovespa/advanced-dh-on-ADHM9/cotacao" title="Cotação da Advanced Digital Health Medicina Preventiva SA (ADHM9)">ADVANCED-DH ON</a></td>,
 <td class="String Column1"><a href="//br.advfn.com/bolsa-de-valores/bovespa/advanced-dh-on-ADHM1/cotacao" title="Cotação da Advanced Digital Health Medicina Preventiva SA (ADHM1)">ADVANCED-DH ON</a></td>,
 <td class="String Column1"><a href="//br.advfn.com/bolsa-de-valores/bovespa/aeris-industria-e-comerc-on-AERI3/cotacao" title="Cotação da Aeris Industria E ComerciodeEquipamentos paraGeracaodeEne

In [32]:
soup.find_all(class_='String Column2 ColumnLast') 
# procura as classes 'String Column2 ColumnLast' e ele mostra tudo que ele encontrou (nesse caso, os tickers) - LISTA

[<td class="String Column2 ColumnLast">ACCC5L</td>,
 <td class="String Column2 ColumnLast">ADHM3</td>,
 <td class="String Column2 ColumnLast">ADHM9</td>,
 <td class="String Column2 ColumnLast">ADHM1</td>,
 <td class="String Column2 ColumnLast">AERI3</td>,
 <td class="String Column2 ColumnLast">AESB3</td>,
 <td class="String Column2 ColumnLast">TIET-DEB43L0</td>,
 <td class="String Column2 ColumnLast">TIET-DEB43B0</td>,
 <td class="String Column2 ColumnLast">TIET-DEB43L1</td>,
 <td class="String Column2 ColumnLast">TIET3</td>,
 <td class="String Column2 ColumnLast">TIET4</td>,
 <td class="String Column2 ColumnLast">AFLT3</td>,
 <td class="String Column2 ColumnLast">AGVE6L</td>,
 <td class="String Column2 ColumnLast">AGVE5L</td>,
 <td class="String Column2 ColumnLast">AGPP5L</td>,
 <td class="String Column2 ColumnLast">AIPN5L</td>,
 <td class="String Column2 ColumnLast">ARSA5L</td>,
 <td class="String Column2 ColumnLast">AGBH5L</td>,
 <td class="String Column2 ColumnLast">BURI5L</td>,
 <

In [33]:
# como acessar o ticker 1? TENHO UMA LISTA
soup.find_all(class_='String Column2 ColumnLast')[0]

<td class="String Column2 ColumnLast">ACCC5L</td>

In [34]:
# primeiro elemento da lista é um objeto e não uma string, por isso, vou usar o método .get_text()
soup.find_all(class_='String Column2 ColumnLast')[0].get_text()

'ACCC5L'

In [36]:
# para o nome da ação, por exemplo acessando o primeiro elemento DA LISTA
soup.find_all(class_='String Column1')[0]

<td class="String Column1"><a href="//br.advfn.com/bolsa-de-valores/bovespa/acaracoacarapna-pna-ACCC5L/cotacao" title="Cotação da ACARACOACARAPNA (ACCC5L)">ACARACOACARAPNA PNA</a></td>

In [37]:
# acessando o nome da ação na posição 10
soup.find_all(class_='String Column1')[10]

<td class="String Column1"><a href="//br.advfn.com/bolsa-de-valores/bovespa/aes-tiete-pn-TIET4/cotacao" title="Cotação da Aes Tiete Energia SA (TIET4)">AES TIETÊ PN</a></td>

In [38]:
# acessando o nome da ação na posição 10 e pegando a informção de texto dentro dela
soup.find_all(class_='String Column1')[10].get_text()

'AES TIETÊ PN'

In [39]:
# usando um for para acessar todas as ações, ou seja, percorrer a lista
for nome in soup.find_all(class_='String Column1'):
    print(nome.get_text())

ACARACOACARAPNA PNA
ADVANCED-DH ON
ADVANCED-DH ON
ADVANCED-DH ON
Aeris Industria E Comerc... ON
AES Brasil Energia ON
AES TIETE E
AES TIETE E
AES TIETE E
AES TIETÊ ON
AES TIETÊ PN
AFLUENTE ON
Agro Industrial De Cerea... PNB
Agro Industrial De Cerea... PNA
Agro Pontal PNA PNA
Agroindustrial e Pastori... PNA
AGROMASSA PNA PNA
Agropecuaria Belo Horizo... PNA
Agropecuaria Buriti PNA
Agropecuaria Ricastro PNA
Agropecuaria Santa Rosa PNA
Agropecuaria Serra Azul PNA
Agropecuaria Vale Rio Acre PNC
Agropecuaria Vale Rio Acre PNA
ALFA ON
ALFA PNB
ALFA PNA
Allied Tecnologia ON
Allpark Empreendimentos ... ON
ALPER ON
ALPER ON
ALPER ON
Alphaville ON
Alphaville ON
Alumazon Componentes Ama... PNA
Aluminio Extrudado Brasi... PN
ALUPAR ON
ALUPAR PN
Amazon Golf PNA
AMBEV S/A ON
Ambipar Participacoes e ... ON
B2W DIGITAL ON
Americanas ON
AMPLA ENERG ON
Andrade Gutierrez Partic... ON
Andrade Gutierrez Partic... ON
Andrade Gutierrez Partic... PN
Andrade Gutierrez Partic... PN
Angelica Pna PNA
ANIMA ON
APIA

In [40]:
for ticker in soup.find_all(class_='String Column2 ColumnLast'):
    print(ticker.get_text())

ACCC5L
ADHM3
ADHM9
ADHM1
AERI3
AESB3
TIET-DEB43L0
TIET-DEB43B0
TIET-DEB43L1
TIET3
TIET4
AFLT3
AGVE6L
AGVE5L
AGPP5L
AIPN5L
ARSA5L
AGBH5L
BURI5L
RCRO5L
AGSR5L
AGSA5L
VLRA7L
VLRA5L
RPAD3
RPAD6
RPAD5
ALLD3
ALPK3
APER1
APER9
APER3
AVLL3
AVLL11
ALMZ5L
ALMN4L
ALUP3
ALUP4
AZGT5L
ABEV3
AMBP3
BTOW3
AMER3
CBEE3
CANT3B
CANT3
CANT4
CANT4B
SAIS5L
ANIM3
APAK5L
AHEV6L
ARZZ3
ARTR-DEB53L0
ARTR-DEB53L1
ARTR-DEB53B0
ARTR-DEB530B
ATMP3
ATOM3
AVIP5L
AZEV4
AZEV3
AZEV2
AZEV1
AZUL4


In [41]:
# armazenar esses dados em 2 listas, por exemplo
acao = []
codigo = []
for nome, ticker in zip( soup.find_all(class_='String Column1'), soup.find_all(class_='String Column2 ColumnLast')):
    acao.append(nome.get_text())
    codigo.append(ticker.get_text())
    

In [42]:
len(acao) , len(codigo)

(65, 65)

In [43]:
# Essas 65 ações e 65 códigos são da primeira página como faço para pegar de outras páginas tb?

# Pode pegar os links dentro da class "odd select first" ou soup.find_all(class_="link_list atoz_list")[0].get_text()

# criar os links para buscar os dados

# https://br.advfn.com/bolsa-de-valores/bovespa/A'
# https://br.advfn.com/bolsa-de-valores/bovespa/B'
# https://br.advfn.com/bolsa-de-valores/bovespa/C'
# https://br.advfn.com/bolsa-de-valores/bovespa/D'
# https://br.advfn.com/bolsa-de-valores/bovespa/E'
# https://br.advfn.com/bolsa-de-valores/bovespa/F'
# https://br.advfn.com/bolsa-de-valores/bovespa/G'

# https://br.advfn.com/bolsa-de-valores/bovespa/X 

import string # letras do alfabeto

#lista de letras do alfabeto
list(string.ascii_uppercase)


['A',
 'B',
 'C',
 'D',
 'E',
 'F',
 'G',
 'H',
 'I',
 'J',
 'K',
 'L',
 'M',
 'N',
 'O',
 'P',
 'Q',
 'R',
 'S',
 'T',
 'U',
 'V',
 'W',
 'X',
 'Y',
 'Z']

In [49]:
 # montando o primeiro link
'https://br.advfn.com/bolsa-de-valores/bovespa/' + list(string.ascii_uppercase)[0]

'https://br.advfn.com/bolsa-de-valores/bovespa/A'

In [52]:
paginas = []
for item in list(string.ascii_uppercase):
    paginas.append('https://br.advfn.com/bolsa-de-valores/bovespa/' + item)

for i in range(10):
    paginas.append('https://br.advfn.com/bolsa-de-valores/bovespa/' + str(i))


In [53]:
# páginas para acessar
paginas

['https://br.advfn.com/bolsa-de-valores/bovespa/A',
 'https://br.advfn.com/bolsa-de-valores/bovespa/B',
 'https://br.advfn.com/bolsa-de-valores/bovespa/C',
 'https://br.advfn.com/bolsa-de-valores/bovespa/D',
 'https://br.advfn.com/bolsa-de-valores/bovespa/E',
 'https://br.advfn.com/bolsa-de-valores/bovespa/F',
 'https://br.advfn.com/bolsa-de-valores/bovespa/G',
 'https://br.advfn.com/bolsa-de-valores/bovespa/H',
 'https://br.advfn.com/bolsa-de-valores/bovespa/I',
 'https://br.advfn.com/bolsa-de-valores/bovespa/J',
 'https://br.advfn.com/bolsa-de-valores/bovespa/K',
 'https://br.advfn.com/bolsa-de-valores/bovespa/L',
 'https://br.advfn.com/bolsa-de-valores/bovespa/M',
 'https://br.advfn.com/bolsa-de-valores/bovespa/N',
 'https://br.advfn.com/bolsa-de-valores/bovespa/O',
 'https://br.advfn.com/bolsa-de-valores/bovespa/P',
 'https://br.advfn.com/bolsa-de-valores/bovespa/Q',
 'https://br.advfn.com/bolsa-de-valores/bovespa/R',
 'https://br.advfn.com/bolsa-de-valores/bovespa/S',
 'https://br

In [58]:
# criar uma função para buscar info em cada página:

def pagina_ativos(url):
    page = requests.get(url)
    soup = BeautifulSoup(page.text, 'html.parser')
    acao = []
    codigo = []
    for nome, ticker in zip( soup.find_all(class_='String Column1'), soup.find_all(class_='String Column2 ColumnLast')):
        acao.append(nome.get_text())
        codigo.append(ticker.get_text())
    
    return acao, codigo
    

In [59]:
# variável paginas tem a lista de todos os links

# testando a função:
pagina_ativos(paginas[0])

(['ACARACOACARAPNA PNA',
  'ADVANCED-DH ON',
  'ADVANCED-DH ON',
  'ADVANCED-DH ON',
  'Aeris Industria E Comerc... ON',
  'AES Brasil Energia ON',
  'AES TIETE E',
  'AES TIETE E',
  'AES TIETE E',
  'AES TIETÊ ON',
  'AES TIETÊ PN',
  'AFLUENTE ON',
  'Agro Industrial De Cerea... PNB',
  'Agro Industrial De Cerea... PNA',
  'Agro Pontal PNA PNA',
  'Agroindustrial e Pastori... PNA',
  'AGROMASSA PNA PNA',
  'Agropecuaria Belo Horizo... PNA',
  'Agropecuaria Buriti PNA',
  'Agropecuaria Ricastro PNA',
  'Agropecuaria Santa Rosa PNA',
  'Agropecuaria Serra Azul PNA',
  'Agropecuaria Vale Rio Acre PNC',
  'Agropecuaria Vale Rio Acre PNA',
  'ALFA ON',
  'ALFA PNB',
  'ALFA PNA',
  'Allied Tecnologia ON',
  'Allpark Empreendimentos ... ON',
  'ALPER ON',
  'ALPER ON',
  'ALPER ON',
  'Alphaville ON',
  'Alphaville ON',
  'Alumazon Componentes Ama... PNA',
  'Aluminio Extrudado Brasi... PN',
  'ALUPAR ON',
  'ALUPAR PN',
  'Amazon Golf PNA',
  'AMBEV S/A ON',
  'Ambipar Participacoes e ..

In [60]:
# rodar para todas as páginas
lst_page_acao = []
lst_page_codigo = []

for url in paginas:
    print(url)
    acao, codigo = pagina_ativos(url)
    lst_page_acao.extend(acao)
    lst_page_codigo.extend(codigo)
    
    print(len(lst_page_acao), len(lst_page_codigo))

    

https://br.advfn.com/bolsa-de-valores/bovespa/A
65 65
https://br.advfn.com/bolsa-de-valores/bovespa/B
171 171
https://br.advfn.com/bolsa-de-valores/bovespa/C
336 336
https://br.advfn.com/bolsa-de-valores/bovespa/D
355 355
https://br.advfn.com/bolsa-de-valores/bovespa/E
427 427
https://br.advfn.com/bolsa-de-valores/bovespa/F
496 496
https://br.advfn.com/bolsa-de-valores/bovespa/G
524 524
https://br.advfn.com/bolsa-de-valores/bovespa/H
540 540
https://br.advfn.com/bolsa-de-valores/bovespa/I
595 595
https://br.advfn.com/bolsa-de-valores/bovespa/J
605 605
https://br.advfn.com/bolsa-de-valores/bovespa/K
612 612
https://br.advfn.com/bolsa-de-valores/bovespa/L
640 640
https://br.advfn.com/bolsa-de-valores/bovespa/M
696 696
https://br.advfn.com/bolsa-de-valores/bovespa/N
711 711
https://br.advfn.com/bolsa-de-valores/bovespa/O
724 724
https://br.advfn.com/bolsa-de-valores/bovespa/P
770 770
https://br.advfn.com/bolsa-de-valores/bovespa/Q
772 772
https://br.advfn.com/bolsa-de-valores/bovespa/R
81

In [62]:
# print(lst_page_acao)
print(lst_page_codigo)

['ACCC5L', 'ADHM3', 'ADHM9', 'ADHM1', 'AERI3', 'AESB3', 'TIET-DEB43L0', 'TIET-DEB43B0', 'TIET-DEB43L1', 'TIET3', 'TIET4', 'AFLT3', 'AGVE6L', 'AGVE5L', 'AGPP5L', 'AIPN5L', 'ARSA5L', 'AGBH5L', 'BURI5L', 'RCRO5L', 'AGSR5L', 'AGSA5L', 'VLRA7L', 'VLRA5L', 'RPAD3', 'RPAD6', 'RPAD5', 'ALLD3', 'ALPK3', 'APER1', 'APER9', 'APER3', 'AVLL3', 'AVLL11', 'ALMZ5L', 'ALMN4L', 'ALUP3', 'ALUP4', 'AZGT5L', 'ABEV3', 'AMBP3', 'BTOW3', 'AMER3', 'CBEE3', 'CANT3B', 'CANT3', 'CANT4', 'CANT4B', 'SAIS5L', 'ANIM3', 'APAK5L', 'AHEV6L', 'ARZZ3', 'ARTR-DEB53L0', 'ARTR-DEB53L1', 'ARTR-DEB53B0', 'ARTR-DEB530B', 'ATMP3', 'ATOM3', 'AVIP5L', 'AZEV4', 'AZEV3', 'AZEV2', 'AZEV1', 'AZUL4', 'BTOW1', 'BTOW9', 'B3SA3', 'BAHI3', 'BAHI9', 'BAHI11', 'BMGB4', 'BPAC12', 'BPAC3', 'BPAC5', 'BPAC9', 'BIDI14', 'BIDI13', 'BIDI10', 'BIDI15', 'BIDI3', 'BIDI9', 'BIDI4', 'BIDI1', 'BIDI16', 'BIDI2', 'BPAN10', 'BPAN4', 'SANB4', 'SANB3', 'BEES4', 'BEES3', 'BDLL3', 'BDLL4', 'BTTL3', 'BALM3', 'BALM4', 'BBML3', 'BBSE3', 'ABCB10', 'ABCB4', 'ABCB10L'

In [66]:
# Bonus 1
# Não foi dado mas em breve vc vão aprender e usar

import pandas as pd

# conda install -c anaconda pandas

df = pd.concat([pd.DataFrame(lst_page_acao, columns=['nome da ação']), 
                pd.DataFrame(lst_page_codigo, columns=['codigo da ação'])], axis=1)
df

nome da ação codigo da ação
0               ACARACOACARAPNA PNA         ACCC5L
1                    ADVANCED-DH ON          ADHM3
2                    ADVANCED-DH ON          ADHM9
3                    ADVANCED-DH ON          ADHM1
4    Aeris Industria E Comerc... ON          AERI3
..                              ...            ...
986                 ESTACIO PART ON          YDUQ3
987                           Zynga         Z2NZ34
988  3R Petroleum Oleo E Gas ... ON          RRRP3
989                3TENTOS ON NM ON          TTEN3
990            524 Participacoes ON         QVQP3B

[991 rows x 2 columns]

In [69]:
# Bonus 2
# Salvar esse dataframe em csv
df.to_csv('C:/Users/Renata/DSdegree/Aula19/webscraping.csv', index=False, encoding='utf-8-sig', sep=';')

In [70]:
# Lembram de outra forma de pegar a lista de links?

# alternativa ->> url antes de entrar na lista A

url = 'https://br.advfn.com/bolsa-de-valores/bovespa'

try:
    page = requests.get(url)
    page.raise_for_status()
    response = page.text #ou page.content
    soup = BeautifulSoup(response, 'html.parser')
    
except Exception as err:
    print('Algo deu errado:', err)

In [71]:
print(soup.prettify())

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html lang="pt-BR" xmlns="http://www.w3.org/1999/xhtml">
 <head>
  <meta content="IE=edge" http-equiv="X-UA-Compatible"/>
  <script type="text/javascript">
   window.USER_COUNTRY = "BR";
window.USER_STATE = "SP";
window.USER_FROM_EU = "";
  </script>
  <!-- APS_HEADER_START -->
  <script type="text/javascript">
   function AdvfnJsQueueLoader(e){this.check=e,this.queue=[],this.id=!1,this.ready=!1;var t=this;this.id=setInterval(function(){t.tick()},50)}AdvfnJsQueueLoader.prototype.processQueue=function(){var e;if(0<this.queue.length)for(e=0;e<this.queue.length;e++)this.queue[e]()},AdvfnJsQueueLoader.prototype.tick=function(){this.check()&&(clearInterval(this.id),this.ready=!0,this.processQueue())},AdvfnJsQueueLoader.prototype.add=function(e){this.ready?e():this.queue.push(e)};
var jqueryReady=new AdvfnJsQueueLoader(function(){return'function'==typeof $});
  </script>
  <script type="text/javascript

In [77]:
# para o primeiro link
soup.find_all('div', id='a-to-z')[0].find_all('a', href=True)[0]['href']

'/bolsa-de-valores/bovespa/A'

In [81]:
# para pegar todos
lista_link = []
for link in soup.find_all('div', id='a-to-z')[0].find_all('a', href=True):
    print('https://br.advfn.com' + link['href'])
    lista_link.append('https://br.advfn.com' + link['href'])

https://br.advfn.com/bolsa-de-valores/bovespa/A
https://br.advfn.com/bolsa-de-valores/bovespa/B
https://br.advfn.com/bolsa-de-valores/bovespa/C
https://br.advfn.com/bolsa-de-valores/bovespa/D
https://br.advfn.com/bolsa-de-valores/bovespa/E
https://br.advfn.com/bolsa-de-valores/bovespa/F
https://br.advfn.com/bolsa-de-valores/bovespa/G
https://br.advfn.com/bolsa-de-valores/bovespa/H
https://br.advfn.com/bolsa-de-valores/bovespa/I
https://br.advfn.com/bolsa-de-valores/bovespa/J
https://br.advfn.com/bolsa-de-valores/bovespa/K
https://br.advfn.com/bolsa-de-valores/bovespa/L
https://br.advfn.com/bolsa-de-valores/bovespa/M
https://br.advfn.com/bolsa-de-valores/bovespa/N
https://br.advfn.com/bolsa-de-valores/bovespa/O
https://br.advfn.com/bolsa-de-valores/bovespa/P
https://br.advfn.com/bolsa-de-valores/bovespa/Q
https://br.advfn.com/bolsa-de-valores/bovespa/R
https://br.advfn.com/bolsa-de-valores/bovespa/S
https://br.advfn.com/bolsa-de-valores/bovespa/T
https://br.advfn.com/bolsa-de-valores/bo